### PR TITLE
Allow passing prepared objects instead of array definition

### PR DESCRIPTION
**Pull request by SilverFire**
Before this pull request I was able to write only

```php
new Server([ 
    'url' => "{protocol}://$host/",
    'variables' => [
        'protocol' => [
            'enum' => ['http', 'https'],
            'default' => 'https', 
        ],
    ],
])
```

It's could be enough, but when it comes to more complex code, I would like to write something like that to be more verbose on what is going on inside my object:

```php
new Server([ 
    'url' => "{protocol}://$host/",
    'variables' => [
        'protocol' => new ServerVariable([
            'enum' => ['http', 'https'],
            'default' => 'https',
        ]),
    ],
])
```

This pull request enhances library so it respects the passed object.

### CODE REVIEW

File: src/spec/Paths.php [hunk 1]
```diff
@@ -39,16 +39,24 @@ class Paths implements SpecObjectInterface, ArrayAccess, Countable, IteratorAggr
...
+            } elseif (is_array($object)) {
                 $this->_paths[$path] = new PathItem($object);
+            } elseif ($object instanceof PathItem) {
+                $this->_paths[$path] = $object;
+            } else {
+                $givenType = gettype($object);
+                if ($givenType === 'object') {
+                    $givenType = get_class($object);
+                }
+                throw new \TypeError(sprintf('Path MUST be either array or PathItem object, "%s" given', $givenType));
```
Comment (cebe) [inline src/spec/Paths.php]: this should throw `TypeErrorException` as annotated in phpdoc

File: src/spec/Responses.php [hunk 2]
```diff
@@ -42,7 +42,9 @@ public function __construct(array $data)
             // From Spec: This field MUST be enclosed in quotation marks (for example, "200") for compatibility between JSON and YAML.
```
Comment (cebe) [inline src/spec/Responses.php]: line 36 above (can't comment it directly) should be `@param Response[]|Reference[]|array`

File: src/spec/Paths.php (hunk 1 above)
Comment (cebe) [reply src/spec/Paths.php]: also instantiation logic in Paths and Responses is very similar but we have the type check with exception here but not in Responses. Imo it should follow same rules.

Comment (cebe) [reply src/spec/Paths.php]: we should probably add it in all three places. i.e. if an object is given, verify that it is correct object as expected in this place.

**Comment by SilverFire** [comment]
Hi, Carsten! I've added a test that covers SpecBaseObject, Paths and PathItem behavior at once.

**Comment by cebe** [comment]
Merged, thank you!

**Pull request merged**